lib/Input.js

```diff
@@ -67,17 +79,71 @@ class Keyboard {
   }
 
   /**
+   * @param {string} keyString
+   * @return {KeyDescription}
+   */
+  _keyDescriptionForString(keyString) {
+    const shift = this._modifiers & 8;
+    const description = {
+      key: '',
+      keyCode: 0,
+      code: '',
+      text: '',
+      location: 0
+    };
+
+    let definition = keyDefinitions[keyString];
+    if (!definition) {
+      console.assert(keyString.length === 1, `Unknown key: "${keyString}"`);
+      definition = {key: keyString};
+    }
+
+    if (definition.key)
+      description.key = definition.key;
+    if (shift && definition.shiftKey)
+      description.key = definition.shiftKey;
+
+    if (definition.keyCode)
+      description.keyCode = definition.keyCode;
+    if (shift && definition.shiftKeyCode)
+      description.keyCode = definition.shiftKeyCode;
+
+    if (definition.code)
+      description.code = definition.code;
+
+    if (definition.location)
+      description.location = definition.location;
+
+    if (description.key.length === 1)
+      description.text = description.key;
+
+    if (definition.text)
+      description.text = definition.text;
+    if (shift && definition.shiftText)
+      description.text = definition.shiftText;
+
+    // if any modifiers besides shift are pressed, no text should be sent
+    if (this._modifiers & ~8)
+      description.text = '';
+
+    return description;
+  }
+
+  /**
    * @param {string} key
    */
   async up(key) {
-    this._modifiers &= ~this._modifierBit(key);
-    this._pressedKeys.delete(key);
+    const description = this._keyDescriptionForString(key);
```

Comment:
this throws if there's no key description as well

Reply:
Done.